### PR TITLE
Do not expand wildcards for datasets of derived variables where not all input variables are available

### LINKED ISSUE #2373: Dataset wildcards may lead to "missing data" errors for derived variables
**Describe the bug**

When using wildcards in a recipe with derived variables, and not all required input variables for the derivation are available, the tool reports this is an error. Example:

```yaml
- {project: CMIP6, dataset: "*", exp: historical, ensemble: r1i1p1f1, grid: "*", institute: "*", timerange: 2000/2005, short_name: swcre, derive: true}
```

will lead to the following error:

```
ERROR   [2109992] Missing data for preprocessor timeseries/swcre:
  - Missing data for Dataset: rsutcs, Amon, CMIP6, FIO-ESM-2-0, CMIP, historical, r1i1p1f1, gn
```

Note that the other input variable  `rsut` exists.

I think the tool should not fail here and simply not consider this missing model at all.

### PR DESCRIPTION
**Pull request by schlunma**
## Description

Do not expand wildcards for datasets of derived variables where not all input variables are available. Currently, this will lead to unexpected missing data errors.

Closes #2373

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] ~All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful~ Codacy issue cannot be solved easily and was already present before that PR here

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/_recipe/recipe.py [hunk 1]
```diff
@@ -133,8 +133,8 @@ def _update_target_grid(dataset, datasets, settings):
     if dataset.facets['dataset'] == grid:
         del settings['regrid']
     elif any(grid == d.facets['dataset'] for d in datasets):
-        representative_ds = _representative_dataset(
-            _select_dataset(grid, datasets))
+        representative_ds = _representative_datasets(
+            _select_dataset(grid, datasets))[0]
```
Comment (valeriupredoi) [inline esmvalcore/_recipe/recipe.py:137]: does that func always return an object of len >=1?

File: esmvalcore/_recipe/to_datasets.py [hunk 2]
```diff
@@ -512,11 +557,11 @@ def _get_input_datasets(dataset: Dataset) -> list[Dataset]:
     return datasets
 
 
-def _representative_dataset(dataset: Dataset) -> Dataset:
-    """Find a representative dataset that has files available."""
+def _representative_datasets(dataset: Dataset) -> list[Dataset]:
+    """Find representative datasets for all input variables."""
     copy = dataset.copy()
```
Comment (valeriupredoi) [inline esmvalcore/_recipe/to_datasets.py:562]: I know it's not you - most prob @bouweandela - but it'd be nice not to use standard Python function names for variables

**Review by valeriupredoi — APPROVED** [review]
looks good to me, cheers, Manu! But as far as I can see, we're avoiding the problem, not fixing it - yet, it's prob not worth going any deeper anyway :beer:

File: esmvalcore/_recipe/recipe.py (hunk 1 above)
Comment (schlunma) [reply esmvalcore/_recipe/recipe.py:137]: I think so. The `[0]` part was previously included within `_representative_dataset`, so if it wasn't a problem then, it's not a problem now 😁

File: esmvalcore/_recipe/to_datasets.py (hunk 2 above)
Comment (schlunma) [reply esmvalcore/_recipe/to_datasets.py:562]: Is `copy` really a standard Python function?

```py
Python 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> copy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'copy' is not defined
```

Comment (valeriupredoi) [reply esmvalcore/_recipe/to_datasets.py:562]: `from copy import deepcopy` - deep learning :rofl:

**Pull request merged**